src/utilities/include/vec.h

```diff
@@ -0,0 +1,318 @@
+    if (&other == this) return *this;
+    if (this->ptr_ != nullptr) {
+      TracyFreeS(this->ptr_, 3);
+      free(this->ptr_);
+    }
+    this->size_ = other.size_;
+    capacity_ = other.size_;
+    auto policy = autoPolicy(this->size_);
+    if (this->size_ != 0) {
+      this->ptr_ = reinterpret_cast<T *>(malloc(this->size_ * sizeof(T)));
+      if (this->ptr_ == nullptr) throw std::bad_alloc();
+      TracyAllocS(ptr_, size_ * sizeof(T), 3);
+      uninitialized_copy(policy, other.begin(), other.end(), this->ptr_);
+    }
+    return *this;
+  }
+
+  Vec<T> &operator=(Vec<T> &&other) {
+    if (&other == this) return *this;
+    if (this->ptr_ != nullptr) {
+      TracyFreeS(ptr_, 3);
+      free(this->ptr_);
+    }
+    this->size_ = other.size_;
+    capacity_ = other.capacity_;
+    this->ptr_ = other.ptr_;
+    other.ptr_ = nullptr;
+    other.size_ = 0;
+    other.capacity_ = 0;
+    return *this;
+  }
+
+  void swap(Vec<T> &other) {
+    std::swap(this->ptr_, other.ptr_);
+    std::swap(this->size_, other.size_);
+    std::swap(capacity_, other.capacity_);
+  }
+
+  inline void push_back(const T &val) {
+    if (this->size_ >= capacity_) {
+      // avoid dangling pointer in case val is a reference of our array
+      T val_copy = val;
+      reserve(capacity_ == 0 ? 128 : capacity_ * 2);
+      this->ptr_[this->size_++] = val_copy;
+      return;
+    }
+    this->ptr_[this->size_++] = val;
+  }
+
+  void reserve(int n) {
+    if (n > capacity_) {
+      T *newBuffer = reinterpret_cast<T *>(malloc(n * sizeof(T)));
+      if (newBuffer == nullptr) throw std::bad_alloc();
+      TracyAllocS(newBuffer, n * sizeof(T), 3);
+      if (this->size_ > 0)
+        uninitialized_copy(autoPolicy(this->size_), this->ptr_,
+                           this->ptr_ + this->size_, newBuffer);
+      if (this->ptr_ != nullptr) {
+        TracyFreeS(ptr_, 3);
+        free(this->ptr_);
+      }
+      this->ptr_ = newBuffer;
+      capacity_ = n;
+    }
+  }
+
+  void resize(int newSize, T val = T()) {
+    bool shrink = this->size_ > 2 * newSize;
+    reserve(newSize);
+    if (this->size_ < newSize) {
+      uninitialized_fill(autoPolicy(newSize - this->size_),
+                         this->ptr_ + this->size_, this->ptr_ + newSize, val);
+    }
+    this->size_ = newSize;
+    if (shrink) shrink_to_fit();
+  }
+
+  void shrink_to_fit() {
+    T *newBuffer = nullptr;
+    if (this->size_ > 0) {
+      newBuffer = reinterpret_cast<T *>(malloc(this->size_ * sizeof(T)));
+      if (newBuffer == nullptr) throw std::bad_alloc();
+      TracyAllocS(newBuffer, size_ * sizeof(T), 3);
+      uninitialized_copy(autoPolicy(this->size_), this->ptr_,
+                         this->ptr_ + this->size_, newBuffer);
+      if (this->ptr_ != nullptr) {
+        TracyFreeS(ptr_, 3);
+        free(this->ptr_);
+      }
+    }
+    this->ptr_ = newBuffer;
+    capacity_ = this->size_;
+  }
+
+  VecView<T> get_view(int offset = 0, int length = -1) {
```

Comment:
How much is get_view used with non-default arguments? Curious if that will all go away with the next refactor? Likewise, what is different about `Vec` and `std::vector` at this point? I recall you said you found a perf difference, but do you know why?

Reply:
it is quite rare, in boolean3 and boolean_result where we need to separate a pq vector into two parts.

differences from std::vector
1. we use parallel fill for large vectors
2. initialization can have uninitialized memory (this is pretty unsafe in general, but we can do this because we will immediately write to them in parallel)
3. we have bound checks (std::vector only perform bound checks if you use at instead of [])

Reply:
Huh, I thought `std::vector` did uninitialized memory too. 1 and 3 sound like reason enough though.